### DISCUSSION TITLE
[RFC] Ableton style Song view

### DISCUSSION THREAD

**Discussion by PaulFreund** [SynthstromAudible/DelugeFirmware#150] [Ideas]
I plan to start with this feature after I finished the extended Keyboard PR but I want to put the idea for the Ableton style song view UX in here so people can give feedback or align if they want to work on something similar or same. Any comments are welcome, especially if I forgot something or something is incompatible with the Song editor in general.

# Integration with current UX
* Ableton style Song view is an alternative to the current row based layout that can be toggled in Community Settings Menu.
* Sections will be repurposed as launch rows.
* Most of the other UX (holding pads of a clip allows editing parameters etc.) should apply as before.

# UX
* Users can set via menu if the instruments (Synth/Kit/MIDI/CV) should be organized vertically or horizontically.
* In horizontal mode 
    * Every column represents an instrument.
    * Every row represents a Section (launch row).
    * The left audition row (Mute/Launch) can be used to arm all clips in that row for start in the next measure.
    * The right audition row (Audition/Section) has no defined function yet (suggestions?).
    * The Left/Right turn knob can be used to scroll to more instruments.
* In Vertical mode
    * Every row represents an instrument.
    * Every column represents a Section (launch column).
    * The bottom row can be used to arm all clips for that column for start in the next measure.
    * The Up/Down turn knob can be used to scroll to more instruments.
    * The right audition row (Audition/Section) has no defined function yet (suggestions?).
* Single clicking on a pad of an instrument will arm it for playing in the next measure. 
* Single clicking on a pad that is already playing it will stop at the next measure.
* "Clip" + Single clicking on a pad will open it in clip view.
* Shift + Single clicking on a pad will immediate launch / stop it
* Holding a clip will allow tuning it's parameters similar as before
* Holding a clip pad and pressing another clip pad will copy all it's contents but will replace the instrument (with preexisting or new one).

# Open questions
* There is still a lot of unused UX that might be useful, ideas welcome!
* Playhead position in sidebar?

# Related topics
* Implement Section play wrap around (start at next available group)
* Launch quantization? 

(v4)

**Comment by weavermedia** [comment]
Alternative idea for clips clicks, including instant launch:
- ~~Shift + Single~~ Double clicking on a pad will open it in clip view.
- Shift + Single clicking on a pad will launch it instantly
- Shift + left audition row (Mute/Launch) to launch all clips in that row instantly (for horizontal mode)

Also, I like the Gate trigger mode for clips in Ableton Live. Pressing the clip pad enables playback, letting go of the pad disables playback. Might be cool for live performance. Trigger mode could be set per track or globally as a main setting.

Question: for vertical mode would the Mute/Launch column still act as mute for that track?

Finally, for clarity to non-Ableton Live users the horizontal mode in this proposal is the style that Live uses.

**Reply by PaulFreund** [reply]
Regarding double click would not work properly since there is no real way of distinguishing properly if it is quick re-(dis)arm. I changed the proposal to use the keyboard key for instant play. Gate trigger is interesting, maybe as a follow up.

The vertical would currently not use the Mute/Audition lady, not yet sure what function could be there.

**Reply by weavermedia** [reply]
(Merged my previous comments into something more coherent here 😀)

Since the keyboard button has no existing association to 'instant launch' but the shift button does it might be better if shift is instant launch and the keyboard button opens the clip in edit view.

Even if the keyboard isn't used to open the clip I still think shift should be used for instant launch, mainly because that's what everyone already knows.

- Shift + clicking on a pad will launch it instantly
- Keyboard + clicking on a pad will open it in clip view

**Reply by PaulFreund** [reply]
You are completely correct that keyboard button is a bad fit. I thought about it again and I think I missed the obvious before:

 Clip+Clicking on the pad

It's pretty much in line with the existing UX with the difference of telling which clip to open. What do you think? I updated my proposal to reflect it. Shift+Pad will again do an immediate start/stop

**Reply by weavermedia** [reply]
I had suggested using the Clip button in one of my deleted comments but I wondered if it might clash with the way we enter/leave this new view. If you use a community setting that would be easy then to use the Clip button.

Personally I like the idea of having this Ableton Live style view as a third option alongside the current Song and Clip views. If that was the case you might end up having this new view enabled/disabled by something like pressing Clip+Song. This means the Clip button won't be usable to enter an individual clip.

What do you think about having this a third view rather than a predetermined choice between the current clip (rows) view and this new view?

**Reply by PaulFreund** [reply]
It will definitely be a community setting even if we use another button for it. It will not clash if I change the behavior for opening clip view from opening on clip button pressed to clip button released. 

I'm not yet sure if it can be harmonized with the current view because of ordering information but everyhing else should be compatible. I'm not yet convinced we need a quick switch for this (community settings are also instant but they require to go in the menu) but I will think about it.

**Comment by weavermedia** [comment]
Launch quantization, like Live, could be a global setting where None is instant and other setting range from 1/16th (or less) to 1 (or more) bar(s).

**Comment by riczikovacs** [comment]
I would love to see this, but I wonder if we could get the best of both worlds by juicing up the horizontal setup by taking some cues from arranger mode. 

Start by changing from all clips on rows to instruments on rows and clips in columns, just like arranger mode. Clips could still reflect their relative length and show playhead position while playing.  

Then show the section a clip belongs to with the color of the clip, clips with no section could be white, just like arranger mode. 

The "Active" clip for a particular instrument could be shown with brightness (just like in the current song mode while arrange mode has control!!), and could be changed with the leftmost clip pad or somethin. I like the idea of holding a clip pad and hitting the clip button to enter it, I often accidentally enter clip edit...

Keep the launch button functionality but assign it to the instrument, like a quantized version of the arranger modes mute buttons. 

Again I would love to try out what you suggest and my take could be missing the point... but one of the cool things about the deluge, to me, is that it is its own thing not too much like ableton...  Though clip launch quantization options (especially gating a clip with 0 quantization aka play an an entire clip like a single synth note!!) would be amazing as well!

**Reply by PaulFreund** [reply]
For now this is out of my scope but I intend to make the implementation flexible so other layouts can be added as well (possibly even an API like for keyboard layouts) that would allow others to make more alternative views. I'm tracking work in #251 now :)
